src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+	Tuplestorestate *tupleStore = NULL;
+	MemoryContext perQueryContext = NULL;
+	MemoryContext oldContext = NULL;
+
+	int attributeCount = 4;
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	DistributedTransactionId *distributedTransctionId = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	perQueryContext = returnSetInfo->econtext->ecxt_per_query_memory;
+
+	oldContext = MemoryContextSwitchTo(perQueryContext);
+
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	returnSetInfo->returnMode = SFRM_Materialize;
+	returnSetInfo->setResult = tupleStore;
+	returnSetInfo->setDesc = tupleDescriptor;
+
+	MemoryContextSwitchTo(oldContext);
+
+	/* MyDistributedTransactionBackend should always be avaliable, just out of paranoia */
+	if (!MyDistributedTransactionBackend)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	distributedTransctionId = GetCurrentDistributedTransctionId();
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	values[0] = ObjectIdGetDatum(MyDatabaseId);
+	values[1] = UInt64GetDatum(distributedTransctionId->initiatorNodeIdentifier);
+	values[2] = UInt64GetDatum(distributedTransctionId->transactionId);
+	values[3] = TimestampTzGetDatum(distributedTransctionId->timestamp);
+
+	tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetAllActiveDistributedTransactions returns all the active distributed
+ * transactions on the shared memory.
+ *
+ * Note that the function does not gurantee a global consistent result, the function
+ * only locks each active backend while copying it. Currently the only caller of this
+ * function is in the src/test folder. The consistency gurantee is sufficient for
+ * testing.
+ */
+List *
+GetAllActiveDistributedTransactions(void)
+{
+	int sessionIndex = 0;
+	List *activeTransactionList = NIL;
+
+	for (sessionIndex = 0; sessionIndex < MaxBackends; ++sessionIndex)
+	{
+		DistributedTransactionBackendData *currentBackend =
+			&distributedTransactionShmemData->sessions[sessionIndex];
+
+		if (currentBackend && currentBackend->transactionId.transactionId != 0)
```

Comment:
How can `curBackend` be NULL here? Given this is a pre-allocated array of the right size that ought to be impossible.